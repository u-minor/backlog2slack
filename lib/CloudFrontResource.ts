import { Fn, Stack } from 'aws-cdk-lib';
import { ICertificate } from 'aws-cdk-lib/aws-certificatemanager';
import {
  AllowedMethods,
  CachePolicy,
  Distribution,
  OriginProtocolPolicy,
  OriginRequestPolicy,
  OriginSslPolicy,
  PriceClass,
  ResponseHeadersPolicy,
  ViewerProtocolPolicy,
} from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';

export interface CloudFrontResourceProps {
  certificate: ICertificate;
  domainName: string;
  lambda: {
    functionUrl: string;
  };
}

export class CloudFrontResource {
  distribution: Distribution;

  constructor(stack: Stack, props: CloudFrontResourceProps) {
    this.distribution = new Distribution(stack, 'Distribution', {
      certificate: props.certificate,
      comment: 'Distribution for backlog2slack',
      domainNames: [props.domainName],
      defaultBehavior: {
        allowedMethods: AllowedMethods.ALLOW_ALL,
        cachePolicy: CachePolicy.CACHING_DISABLED,
        origin: new HttpOrigin(
          Fn.select(2, Fn.split('/', props.lambda.functionUrl)),
          {
            protocolPolicy: OriginProtocolPolicy.HTTPS_ONLY,
            originSslProtocols: [OriginSslPolicy.TLS_V1_2],
          },
        ),
        originRequestPolicy: OriginRequestPolicy.fromOriginRequestPolicyId(
          stack,
          'OriginRequestPolicy',
          '88a5eaf4-2fd4-4709-b370-b4c650ea3fcf',
        ), // https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/using-managed-origin-request-policies.html#managed-origin-request-policy-cors-s3
        responseHeadersPolicy:
          ResponseHeadersPolicy.fromResponseHeadersPolicyId(
            stack,
            'ResponseHeadersPolicy',
            '5cc3b908-e619-4b99-88e5-2cf7f45965bd',
          ), // https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/using-managed-response-headers-policies.html#managed-response-headers-policies-cors-preflight
        viewerProtocolPolicy: ViewerProtocolPolicy.HTTPS_ONLY,
      },
      priceClass: PriceClass.PRICE_CLASS_200,
    });
  }
}
