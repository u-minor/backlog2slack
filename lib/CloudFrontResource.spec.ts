import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { CloudFrontResource } from './CloudFrontResource';
import { Certificate } from 'aws-cdk-lib/aws-certificatemanager';

describe('CloudFrontResource', () => {
  let template: Template;

  beforeAll(() => {
    const app = new App();
    const stack = new Stack(app, 'TestStack');
    const certificate = Certificate.fromCertificateArn(
      new Stack(),
      'TestCertificate',
      'arn:aws:acm:us-east-1:1234567890:certificate/DUMMY_ID',
    );

    new CloudFrontResource(stack, {
      certificate,
      domainName: 'test.example.com',
      lambda: {
        functionUrl: 'https://RANDOM_ID.lambda-url.YOUR_REGION.on.aws/',
      },
    });

    template = Template.fromStack(stack);
  });

  it('should generate valid Cloudfront distribution resource', () => {
    const actual = template.findResources('AWS::CloudFront::Distribution');

    expect(actual).toMatchSnapshot();
  });
});
