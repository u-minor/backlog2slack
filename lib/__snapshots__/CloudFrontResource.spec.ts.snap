// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloudFrontResource should generate valid Cloudfront distribution resource 1`] = `
{
  "Distribution830FAC52": {
    "Properties": {
      "DistributionConfig": {
        "Aliases": [
          "test.example.com",
        ],
        "Comment": "Distribution for backlog2slack",
        "DefaultCacheBehavior": {
          "AllowedMethods": [
            "GET",
            "HEAD",
            "OPTIONS",
            "PUT",
            "PATCH",
            "POST",
            "DELETE",
          ],
          "CachePolicyId": "4135ea2d-6df8-44a3-9df3-4b5a84be39ad",
          "Compress": true,
          "OriginRequestPolicyId": "88a5eaf4-2fd4-4709-b370-b4c650ea3fcf",
          "ResponseHeadersPolicyId": "5cc3b908-e619-4b99-88e5-2cf7f45965bd",
          "TargetOriginId": "TestStackDistributionOrigin1B0A090BF",
          "ViewerProtocolPolicy": "https-only",
        },
        "Enabled": true,
        "HttpVersion": "http2",
        "IPV6Enabled": true,
        "Origins": [
          {
            "CustomOriginConfig": {
              "OriginProtocolPolicy": "https-only",
              "OriginSSLProtocols": [
                "TLSv1.2",
              ],
            },
            "DomainName": "RANDOM_ID.lambda-url.YOUR_REGION.on.aws",
            "Id": "TestStackDistributionOrigin1B0A090BF",
          },
        ],
        "PriceClass": "PriceClass_200",
        "ViewerCertificate": {
          "AcmCertificateArn": "arn:aws:acm:us-east-1:1234567890:certificate/DUMMY_ID",
          "MinimumProtocolVersion": "TLSv1.2_2021",
          "SslSupportMethod": "sni-only",
        },
      },
    },
    "Type": "AWS::CloudFront::Distribution",
  },
}
`;
